# EDA & Preprocessing for Student Readiness Classification

In this notebook, we will:
1. **Explore** the ASSISTments dataset  
2. **Clean** and handle missing values  
3. **Engineer** features for an early‑interaction “readiness” label  
4. **Transform** data via pipelines  
5. **Split** into train/test sets  

---

## 1. Import Libraries

We import **pandas** and **numpy** for data handling, and **scikit‑learn** modules for preprocessing and splitting.

In [74]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

---

## 2. Load the Dataset

Load the full ASSISTments CSV into a DataFrame.  

In [75]:
df = pd.read_csv('assistments_2009_2010.csv')

---

## 3. Initial Exploration

- **Shape** tells us total rows & columns.  
- **head()** previews the first 5 rows.  
- **info()** shows dtypes & non‑null counts.  
- **isnull().sum()** reveals missing values per column.

In [76]:
print("Shape:", df.shape)
display(df.head())
df.info()
display(df.isnull().sum())

Shape: (1011079, 20)


order_id  assignment_id  user_id  assistment_id  problem_id  original  \
0  20224085         232368    73963          42904       76429         0   
1  20224095         232368    73963          42904       76430         0   
2  20224113         232368    73963          42904       76431         0   
3  20224123         232368    73963          42904       76432         0   
4  20224142         232368    73963          42904       76433         0   

   correct  attempt_count  ms_first_response_time tutor_mode answer_type  \
0      0.0              3                106016.0      tutor    choose_1   
1      1.0              1                194187.0      tutor    choose_1   
2      1.0              1                 12734.0      tutor     algebra   
3      1.0              1                333484.0      tutor    choose_1   
4      0.0              2                 52828.0      tutor     algebra   

   sequence_id  student_class_id  position problem_set_type  base_sequence_id  \
0       6272.0           11816.0      93.0   MasterySection            6272.0   
1       6272.0           11816.0      93.0   MasterySection            6272.0   
2       6272.0           11816.0      93.0   MasterySection            6272.0   
3       6272.0           11816.0      93.0   MasterySection            6272.0   
4       6272.0           11816.0      93.0   MasterySection            6272.0   

  list_skill_ids list_skills  teacher_id  school_id  
0            NaN         NaN     22763.0       73.0  
1            NaN         NaN     22763.0       73.0  
2            NaN         NaN     22763.0       73.0  
3            NaN         NaN     22763.0       73.0  
4            NaN         NaN     22763.0       73.0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1011079 entries, 0 to 1011078
Data columns (total 20 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   order_id                1011079 non-null  int64  
 1   assignment_id           1011079 non-null  int64  
 2   user_id                 1011079 non-null  int64  
 3   assistment_id           1011079 non-null  int64  
 4   problem_id              1011079 non-null  int64  
 5   original                1011079 non-null  int64  
 6   correct                 1011079 non-null  float64
 7   attempt_count           1011079 non-null  int64  
 8   ms_first_response_time  1010954 non-null  float64
 9   tutor_mode              1011079 non-null  object 
 10  answer_type             1011079 non-null  object 
 11  sequence_id             1000073 non-null  float64
 12  student_class_id        1000073 non-null  float64
 13  position                1000073 non-null  float64
 14  pr

order_id                       0
assignment_id                  0
user_id                        0
assistment_id                  0
problem_id                     0
original                       0
correct                        0
attempt_count                  0
ms_first_response_time       125
tutor_mode                     0
answer_type                    0
sequence_id                11006
student_class_id           11006
position                   11006
problem_set_type           11006
base_sequence_id           11006
list_skill_ids            577989
list_skills               577989
teacher_id                 11006
school_id                  28919
dtype: int64

In [77]:
# Map answer types to broader categories
answer_type_mapping = {
    'algebra': 'Problem-Solving',
    'choose_1': 'Knowledge-Based',
    'fill_in_1': 'Knowledge-Based',
    'choose_n': 'Strategic Reasoning',
    'open_response': 'Problem-Solving',
    'rank': 'Strategic Reasoning',
    'external': 'Knowledge-Based'  # assuming "external" is simple lookup questions
}

# Apply the mapping
df['answer_type_grouped'] = df['answer_type'].map(answer_type_mapping)
df['answer_type_grouped'].value_counts()

answer_type_grouped
Problem-Solving        580943
Knowledge-Based        423038
Strategic Reasoning      7098
Name: count, dtype: int64

---

## 4. Handle Missing Values

- We fill **ms_first_response_time** with its median (only ~125 missing).  
- We drop columns with large null counts or that aren’t needed for our early‑interaction model.  

In [78]:
# 4.1 Impute ms_first_response_time
median_time = df['ms_first_response_time'].median()
df['ms_first_response_time'] = df['ms_first_response_time'].fillna(median_time)

# 4.2 Drop columns with excessive or non‑essential nulls
cols_to_drop = [
    'sequence_id', 'student_class_id', 'position', 'problem_set_type',
    'base_sequence_id', 'teacher_id', 'school_id',
    'list_skill_ids', 'list_skills'
]
df = df.drop(columns=cols_to_drop)

---

## 5. Feature Engineering: First Interaction

We extract **each student’s very first problem interaction**. This lets us classify readiness after just one event.  

In [79]:
# Sort by student and time, then select their first event
df_sorted = df.sort_values(['user_id', 'order_id'])
first_interactions = df_sorted.groupby('user_id').first().reset_index()

---

## 6. Define Readiness Categories

We label each student into:
1. **Fundamentals** (Category 1)  
2. **More practice** (Category 2)  
3. **Advanced** (Category 3)  
based on correctness, attempts, and response time.

In [80]:
# Determine a “fast/slow” threshold (median response time)
threshold_time = first_interactions['ms_first_response_time'].median()

def assign_category(row):
    if (row['correct'] == 0) or (row['attempt_count'] > 2) or (row['ms_first_response_time'] > threshold_time):
        return 1  # Needs reinforcement of fundamentals
    elif (row['correct'] == 1) and (row['attempt_count'] == 1) and (row['ms_first_response_time'] <= threshold_time):
        return 3  # Advanced level
    else:
        return 2  # Needs more practice

first_interactions['category'] = first_interactions.apply(assign_category, axis=1)

---

## 7. Select Features & Labels

We’ll use **attempt_count**, **response time**, and two categorical columns to predict the category.  

In [81]:
feature_cols = ['attempt_count', 'ms_first_response_time', 'tutor_mode', 'answer_type_grouped']
X = first_interactions[feature_cols]
y = first_interactions['category']

---

## 8. Build Preprocessing Pipelines

We create separate transformers for numeric vs categorical data, then merge via `ColumnTransformer`.

In [82]:
# Numeric pipeline: median impute + standard scaling
numeric_features = ['attempt_count', 'ms_first_response_time']
numeric_transformer = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

# Categorical pipeline: constant impute + one‑hot encode
categorical_features = ['tutor_mode', 'answer_type_grouped']
categorical_transformer = Pipeline([
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

# Combine them
preprocessor = ColumnTransformer([
    ('num', numeric_transformer, numeric_features),
    ('cat', categorical_transformer, categorical_features)
])

---

## 9. Apply Transformations & Split

- **fit_transform** applies all imputations, scalings, and encodings.  
- We then split data for modeling (80% train, 20% test).

In [83]:
# Transform features
X_processed = preprocessor.fit_transform(X)

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(
    X_processed, y, test_size=0.1, random_state=42, stratify=y
)

# Saving the split
np.savez('data_split.npz', X_train=X_train, X_test=X_test, y_train=y_train, y_test=y_test)

---

# Next Steps

1. **Train** a classifier (e.g., RandomForest, XGBoost).  
2. **Evaluate** with accuracy, confusion matrix, etc.  
3. **Iterate**: add more features, tune thresholds, or try cross‑validation.

---

# Compressing Original Data

For preparing to upload to github

In [88]:
import gzip
import shutil

# Path to the original CSV file
input_file = 'assistments_2009_2010.csv'
# Path to the compressed output file
output_file = 'assistments_2009_2010.csv.gz'

# Compress the CSV file
with open(input_file, 'rb') as f_in:
    with gzip.open(output_file, 'wb') as f_out:
        shutil.copyfileobj(f_in, f_out)

print(f"File '{input_file}' has been compressed to '{output_file}'")

File 'assistments_2009_2010.csv' has been compressed to 'assistments_2009_2010.csv.gz'
